### ISSUE TITLE
lxc launch fails: Too many open files (regression from 19.09)

### ISSUE THREAD

**Issue by andersk** [NixOS/nixpkgs#75467]
**Describe the bug**
`lxc launch ubuntu:18.04` fails on NixOS unstable (20.03) with an error like this:

`Creating the container`
`Error: Failed instance creation: Create container from image: Failed to create image volume: Unpack failed, Failed to run: unsquashfs -f -d /var/lib/lxd/storage-pools/default/images/f75468c572cc50eca7f76391182e6fdaf58431f84c3d35a2c92e83814e701698_tmp/rootfs -n /var/lib/lxd/images/f75468c572cc50eca7f76391182e6fdaf58431f84c3d35a2c92e83814e701698.rootfs: FATAL ERROR:write_file: failed to create file /var/lib/lxd/storage-pools/default/images/f75468c572cc50eca7f76391182e6fdaf58431f84c3d35a2c92e83814e701698_tmp/rootfs/usr/lib/klibc/bin/run-init, because Too many open files.`

This worked in NixOS 19.09.

**To Reproduce**
This can be reproduced on the nixos-minimal live CD. I checked that nixos-minimal-19.09.1548.3a1861fcabc-x86_64-linux.iso works and nixos-minimal-20.03pre204199.3140fa89c51-x86_64-linux.iso fails.

1. Configure `virtualisation.lxd.enable = true;` in `/etc/nixos/configuration.nix`, and `sudo nixos-rebuild switch`
2. `sudo lxd init --auto`
3. `sudo lxc launch ubuntu:18.04`

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.1, NixOS, 20.03.git.18f39bf (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.1`
 - channels(root): `"nixos-20.03pre204199.3140fa89c51"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: [lxd]
# a list of nixos modules affected by the problem
module: [virtualisation.lxd]
```

cc: @fpletz @wucke13

**Comment by primeos** [comment]
I also had this issue and fixed it by adding `LimitNOFILE = 65536;` to the systemd unit for LXD. E.g. something like this (maybe someone has time to make a PR, I'm not using the official module, but it shouldn't make a difference):
```nix
      serviceConfig = {
        ExecStart = "@${pkgs.lxd.bin}/bin/lxd lxd --group lxd";
        Type = "simple";
        KillMode = "process"; # when stopping, leave the containers alone
        LimitNOFILE = 65536;
      };
```

The upstream recommendation is even higher (1048576): https://github.com/lxc/lxd/blob/master/doc/production-setup.md

**Comment by megheaiulian** [comment]
This would be fixed by using the option in #71978

**Comment by andersk** [comment]
Still doesn’t work with #71978 and `virtualisation.lxd.recommendedSysctlSettings = true`.

```console
$ lxc launch ubuntu:18.04
Creating the container
Container name is: right-macaque
Starting right-macaque
Error: Failed to run: /nix/store/cknsj5caig0wkr8imm8vhg3rgd0fha9a-lxd-3.18-bin/bin/.lxd-wrapped forkstart right-macaque /var/lib/lxd/containers /var/log/lxd/right-macaque/lxc.conf:
Try `lxc info --show-log local:right-macaque` for more info

$ lxc info --show-log local:right-macaque
Name: right-macaque
Location: none
Remote: unix://
Architecture: x86_64
Created: 2019/12/27 02:30 UTC
Status: Stopped
Type: persistent
Profiles: default

Log:

lxc right-macaque 20191227023032.125 WARN     initutils - initutils.c:setproctitle:341 - Invalid argument - Failed to set cmdline
lxc right-macaque 20191227023032.392 ERROR    cgfsng - cgroups/cgfsng.c:__do_cgroup_enter:1500 - No space left on device - Failed to enter cgroup "/sys/fs/cgroup/cpuset//lxc.monitor/right-macaque/cgroup.procs"
lxc right-macaque 20191227023032.392 ERROR    start - start.c:__lxc_start:2009 - Failed to enter monitor cgroup
lxc right-macaque 20191227023032.393 ERROR    lxccontainer - lxccontainer.c:wait_on_daemonized_start:873 - Received container state "STOPPING" instead of "RUNNING"
lxc right-macaque 20191227023032.510 WARN     cgfsng - cgroups/cgfsng.c:cgfsng_monitor_destroy:1180 - No space left on device - Failed to move monitor 2129 to "/sys/fs/cgroup/cpuset//lxc.pivot/cgroup.procs"

lxc 20191227023032.511 WARN     commands - commands.c:lxc_cmd_rsp_recv:135 - Connection reset by peer - Failed to receive response for command "get_state"
```

**Comment by megheaiulian** [comment]
This is a bug in lxc. This PR https://github.com/lxc/lxc/pull/3109 seems to fix it but there is no release with it yet. Currently I am using a patch and it seems to work properly: https://github.com/plumelo/nixos/blob/master/overlays/lxd/default.nix#L26

**Comment by andersk** [comment]
I assume this was fixed in #83337.
